zkir/Dialect/Field/Conversions/FieldToModArith/Extension/QuadraticExtensionField.cpp

```diff
@@ -15,12 +15,89 @@ limitations under the License.
 
 #include "zkir/Dialect/Field/Conversions/FieldToModArith/Extension/QuadraticExtensionField.h"
 
+#include "zkir/Dialect/Field/Conversions/FieldToModArith/ConversionUtils.h"
+#include "zkir/Dialect/ModArith/IR/ModArithOps.h"
+
 namespace mlir::zkir::field {
 
-Value QuadraticExtensionField::square(Value v) { return nullptr; }
+Value QuadraticExtensionField::square(Value v) {
+  // For v = c0 + c1 * u where u² = β:
+
+  auto coeffs = toCoeffs(b, v);
+  auto c0 = coeffs[0];
+  auto c1 = coeffs[1];
+
+  // v0 = c0 - c1
+  Value v0 = b.create<mod_arith::SubOp>(c0, c1);
+
+  // v1 = c0 - βc1
+  auto betaC1 = b.create<mod_arith::MulOp>(nonResidue, c1);
+  auto v1 = b.create<mod_arith::SubOp>(c0, betaC1);
+
+  // v2 = c0 * c1
+  auto v2 = b.create<mod_arith::MulOp>(c0, c1);
+
+  // v0 = v0 * v1 + v2
+  auto v0TimesV1 = b.create<mod_arith::MulOp>(v0, v1);
+  v0 = b.create<mod_arith::AddOp>(v0TimesV1, v2);
+
+  // r1 = 2 * v2
+  auto r1 = b.create<mod_arith::DoubleOp>(v2);
+  // r0 = v0 + βv2
+  auto betaV2 = b.create<mod_arith::MulOp>(nonResidue, v2);
+  auto r0 = b.create<mod_arith::AddOp>(v0, betaV2);
+  return fromCoeffs(b, type, {r0, r1});
+}
+
+Value QuadraticExtensionField::mul(Value x, Value y) {
+  // For x = x0 + x1 * u and y = y0 + y1 * u where u² = β:
```

Comment:
can we use subscript like before?

Reply:
I'd vote for subscript since it's more readable and we already use subscript in a lot of place.